app/controllers/incomeSources/add/YouCannotGoBackErrorController.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.incomeSources.add
+
+import auth.MtdItUser
+import config.{AgentItvcErrorHandler, FrontendAppConfig, ItvcErrorHandler}
+import controllers.agent.predicates.ClientConfirmedController
+import controllers.predicates._
+import enums.IncomeSourceJourney.IncomeSourceType
+import enums.JourneyType.{Add, JourneyType}
+import play.api.mvc._
+import services.{CreateBusinessDetailsService, IncomeSourceDetailsService, SessionService}
+import uk.gov.hmrc.auth.core.AuthorisedFunctions
+import utils.IncomeSourcesUtils
+import views.html.incomeSources.add.{IncomeSourceNotAddedError, YouCannotGoBackError}
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class YouCannotGoBackErrorController @Inject()(val checkSessionTimeout: SessionTimeoutPredicate,
+                                               val authenticate: AuthenticationPredicate,
+                                               val authorisedFunctions: AuthorisedFunctions,
+                                               val retrieveNinoWithIncomeSources: IncomeSourceDetailsPredicate,
+                                               val incomeSourceDetailsService: IncomeSourceDetailsService,
+                                               val retrieveBtaNavBar: NavBarPredicate,
+                                               val cannotGoBackError: YouCannotGoBackError)
+                                              (implicit val appConfig: FrontendAppConfig,
+                                               mcc: MessagesControllerComponents,
+                                               val ec: ExecutionContext,
+                                               val itvcErrorHandler: ItvcErrorHandler,
+                                               val itvcErrorHandlerAgent: AgentItvcErrorHandler) extends ClientConfirmedController with IncomeSourcesUtils {
+
+
+  def handleRequest(isAgent: Boolean, incomeSourceType: IncomeSourceType)
+                   (implicit user: MtdItUser[_]): Future[Result] = withIncomeSourcesFS {
+    Future.successful(Ok(cannotGoBackError(isAgent, incomeSourceType)))
+  }
+
+  def show(incomeSourceType: IncomeSourceType): Action[AnyContent] = (checkSessionTimeout andThen authenticate
```

Comment:
if we can add withIncomeSourcesFS check on these controllers as well

Reply:
In this controller there's withIncomeSourcesFS on the handleRequest. If I do the version that checks for the session service value, then this page will infinitely redirect to itself. I'll go and add it to the other error pages though if you think that's worthwhile